### PR TITLE
Move templates from AbstractAdmin and Pool to TemplateRegistry

### PR DESCRIPTION
**Pull request by TimoBakx**
I am targeting this branch, because these changes are backwards compatible.

Related to #3947

## Changelog

```markdown
### Added
- Added `TemplateRegistry`, `TemplateRegistryInterface` and `MutableTemplateRegistryInterface` to handle all template registration related functionality from both `AbstractAdmin` and `Pool`.

### Deprecated
- Deprecated `AbstractAdmin` methods `getTemplate` and `getTemplates`.
- Deprecated `AbstractAdmin` attribute `$templates`.
- Deprecated `Pool` methods `getTemplate`, `setTemplates` and `getTemplates`.
- Deprecated `Pool` attribute `$templates`.
- Deprecated Twig function `get_pool_template`.
```

## To do

- [X] Update the tests
- [x] Update the documentation
- [X] Add an upgrade note
- [x] Fix BC in `Pool`
- [x] Add `@author` tags
- [x] Fix `@deprecated` tags (prefix with "since 3.x, will be dropped in 4.0.")

## Subject

This PR is part of a series of PRs to eventually move all admin template related functionality to a separate component.

This PR itself moves all template registration functionality from `AbstractAdmin` and `Pool` and moves it into the class `TemplateRegistry` and separate services (one per admin and a general one).

## Next steps

- After this PR, a follow-up PR should be made to use a separate Service Locator instead of the general container to find the correct admin template registries. This needs to have the dependency on Symfony DependencyInjection to be set to 3.3 or higher (currently `^2.8 || ^3.2 || ^4.0`). I will also look at other places where a Service Locator could be used instead of the general container.
- After that, a way to set admin templates through the configuration should be added (to replace the `setTemplate` and `setTemplates` method calls defined within the Admin service definition) and deprecate the `AbstractAdmin::setTemplate()` and `AbstractAdmin::setTemplates()` methods.

### CODE REVIEW

**Comment by TimoBakx** [comment]
I will update/add documentation after the code reviews. I need to make sure I'm working in the right direction first.

File: src/Admin/AbstractAdmin.php [hunk 1]
```diff
@@ -1122,33 +1127,52 @@ public function generateMenuUrl($name, array $parameters = [], $absolute = Routi
...
+    }
+
+    /**
+     * @deprecated Use TemplateRegistry services instead
+     *
+     * @param array $templates
+     */
     public function setTemplates(array $templates)
     {
-        $this->templates = $templates;
```
Comment (greg0ire) [inline src/Admin/AbstractAdmin.php]: I think this could/should be kept for BC (if an admin accesses that property, it will still be filled properly). WDYT @sonata-project/contributors ?

File: src/Admin/TemplateRegistry.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Admin;
+
+class TemplateRegistry implements TemplateRegistryInterface
```
Comment (greg0ire) [inline src/Admin/TemplateRegistry.php]: Should be final

File: src/Admin/TemplateRegistryInterface.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+     *
+     * @return string
+     */
+    public function getTemplate($name);
+
+    /**
+     * @param string $name
+     * @param string $template
+     */
+    public function setTemplate($name, $template);
```
Comment (greg0ire) [inline src/Admin/TemplateRegistryInterface.php]: I think we could move the setters to a `MutableTemplateRegistryInterface`, and maybe remove them on master? And use a constructor instead?

File: src/Admin/TemplateRegistry.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+{
```
Comment (greg0ire) [inline src/Admin/TemplateRegistry.php]: Please add a constructor

File: src/Block/AdminListBlockService.php [hunk 5]
```diff
@@ -23,16 +24,25 @@
...
+    /**
+     * @var TemplateRegistryInterface
+     */
+    protected $templateRegistry;
+
     /**
      * @param string $name
      */
-    public function __construct($name, EngineInterface $templating, Pool $pool)
+    public function __construct($name, EngineInterface $templating, Pool $pool, TemplateRegistryInterface $templateRegistry)
```
Comment (greg0ire) [inline src/Block/AdminListBlockService.php]: The last argument should be optional

File: src/Block/AdminListBlockService.php [hunk 6]
```diff
@@ -23,16 +24,25 @@
...
 {
+    /**
+     * @var Pool
+     */
     protected $pool;
 
+    /**
+     * @var TemplateRegistryInterface
+     */
+    protected $templateRegistry;
```
Comment (greg0ire) [inline src/Block/AdminListBlockService.php]: Should be private

File: src/Admin/AbstractAdmin.php [hunk 7]
```diff
@@ -400,6 +400,11 @@
      */
     protected $templates = [];
 
+    /**
+     * @var TemplateRegistryInterface
+     */
+    protected $templateRegistry;
```
Comment (greg0ire) [inline src/Admin/AbstractAdmin.php]: Should be private

File: src/Block/AdminListBlockService.php [hunk 8]
```diff
@@ -23,16 +24,25 @@
...
     /**
      * @param string $name
      */
-    public function __construct($name, EngineInterface $templating, Pool $pool)
+    public function __construct($name, EngineInterface $templating, Pool $pool, TemplateRegistryInterface $templateRegistry)
     {
         parent::__construct($name, $templating);
 
         $this->pool = $pool;
+        $this->templateRegistry = $templateRegistry;
```
Comment (greg0ire) [inline src/Block/AdminListBlockService.php]: That would become `$templateRegistry ?: new TemplateRegistry`

File: src/Block/AdminListBlockService.php (hunk 5 above)
Comment (kunicmarko20) [inline src/Block/AdminListBlockService.php]: Are you sure this is not over 120?

File: src/Controller/CRUDController.php [hunk 9]
```diff
@@ -62,6 +63,13 @@ class CRUDController implements ContainerAwareInterface
      */
     protected $admin;
 
+    /**
+     * The template registry of the related Admin class.
+     *
+     * @var TemplateRegistryInterface
+     */
+    protected $templateRegistry;
```
Comment (greg0ire) [inline src/Controller/CRUDController.php]: should be private

File: src/Controller/CoreController.php [hunk 10]
```diff
@@ -140,6 +141,14 @@ protected function getAdminPool()
         return $this->container->get('sonata.admin.pool');
     }
 
+    /**
+     * @return TemplateRegistryInterface
+     */
+    protected function getTemplateRegistry()
```
Comment (greg0ire) [inline src/Controller/CoreController.php]: Should be private or at least, final

File: src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php [hunk 11]
```diff
@@ -326,7 +327,7 @@ public function applyDefaults(ContainerBuilder $container, $serviceId, array $at
 
         $definition->addMethodCall('showMosaicButton', [$showMosaicButton]);
 
-        $this->fixTemplates($container, $definition, isset($overwriteAdminConfiguration['templates']) ? $overwriteAdminConfiguration['templates'] : ['view' => []]);
+        $this->fixTemplates($serviceId, $container, $definition, isset($overwriteAdminConfiguration['templates']) ? $overwriteAdminConfiguration['templates'] : ['view' => []]);
```
Comment (kunicmarko20) [inline src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: over 120.

File: src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php [hunk 12]
```diff
@@ -337,7 +338,10 @@ public function applyDefaults(ContainerBuilder $container, $serviceId, array $at
         return $definition;
     }
 
-    public function fixTemplates(ContainerBuilder $container, Definition $definition, array $overwrittenTemplates = [])
+    /**
+     * @param string $serviceId
+     */
+    public function fixTemplates($serviceId, ContainerBuilder $container, Definition $definition, array $overwrittenTemplates = [])
```
Comment (kunicmarko20) [inline src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: over 120.

File: src/Resources/config/core.xml [hunk 13]
```diff
@@ -71,5 +71,11 @@
         <service id="sonata.admin.twig.global" class="Sonata\AdminBundle\Twig\GlobalVariables" public="true">
             <argument type="service" id="sonata.admin.pool"/>
         </service>
+        <!-- templates -->
+        <service id="sonata.admin.template_registry" class="Sonata\AdminBundle\Admin\TemplateRegistry" public="false">
```
Comment (greg0ire) [inline src/Resources/config/core.xml]: Why does the pool need a template registry by the way? Should this be rename to indicate what particularity this template registry has?

File: src/Twig/Extension/SonataAdminExtension.php [hunk 14]
```diff
@@ -50,7 +51,12 @@ class SonataAdminExtension extends AbstractExtension
      */
     private $xEditableTypeMapping = [];
 
-    public function __construct(Pool $pool, LoggerInterface $logger = null, TranslatorInterface $translator = null)
+    /**
+     * @var TemplateRegistryInterface
+     */
+    private $templateRegistry;
+
+    public function __construct(Pool $pool, LoggerInterface $logger = null, TranslatorInterface $translator = null, TemplateRegistryInterface $templateRegistry = null)
```
Comment (kunicmarko20) [inline src/Twig/Extension/SonataAdminExtension.php]: over 120 characters, please linebreak.

Comment (greg0ire) [inline src/Twig/Extension/SonataAdminExtension.php]: Please use a multiline signature

File: src/Twig/Extension/TemplateRegistryExtension.php [hunk 15]
```diff
@@ -11,21 +11,29 @@
...
+    /**
+     * @var ContainerInterface
+     */
+    private $container;
+
+    public function __construct(TemplateRegistryInterface $generalTemplateRegistry, ContainerInterface $container)
     {
-        $this->pool = $pool;
+        $this->generalTemplateRegistry = $generalTemplateRegistry;
+        $this->container = $container;
```
Comment (greg0ire) [inline src/Twig/Extension/TemplateRegistryExtension.php]: You can and probably should bump the sf dependency to 3.4, to allow a service locator here. We wanted to do it, and were just waiting for a good reason to do so. I think this would be it.

File: src/Admin/AbstractAdmin.php [hunk 16]
```diff
@@ -1122,33 +1127,52 @@ public function generateMenuUrl($name, array $parameters = [], $absolute = Routi
...
 
     /**
+     * @deprecated Use TemplateRegistry services instead
+     *
      * @param string $name
      * @param string $template
      */
     public function setTemplate($name, $template)
     {
-        $this->templates[$name] = $template;
```
Comment (greg0ire) [inline src/Admin/AbstractAdmin.php]: Same here

File: src/Twig/Extension/SonataAdminExtension.php [hunk 17]
```diff
@@ -139,7 +146,7 @@ public function renderListElement(
     ) {
         $template = $this->getTemplate(
             $fieldDescription,
-            $fieldDescription->getAdmin()->getTemplate('base_list_field'),
+            $this->templateRegistry->getTemplate('base_list_field'),
```
Comment (greg0ire) [inline src/Twig/Extension/SonataAdminExtension.php]: I think that's an issue right? How would you know if this is the correct template registry for that particular admin?

**Comment by greg0ire** [comment]
happening

**Comment by greg0ire** [comment]
> This needs to have the dependency on Symfony DependencyInjection to be set to 3.3 or higher (currently ^2.8 || ^3.2 || ^4.0), which is not BC.

We'll consider it BC :P

**Comment by kunicmarko20** [comment]
Won't that break older Symfony apps?

**Comment by greg0ire** [comment]
> Won't that break older Symfony apps?

No, because they won't even be able to upgrade.

**Comment by kunicmarko20** [comment]
So no one using < symfony3.4 won't be able to update? This seems more appropriate for master then.

**Comment by greg0ire** [comment]
> So people using < symfony3.4 won't be able to update SonataAdminBundle? This seems more appropriate for master then.

I think we discussed, allowed and planned this move.

**Comment by kunicmarko20** [comment]
Didn't we just discuss https://github.com/sonata-project/dev-kit/issues/394 dropping it on master and not on stable branches?

**Comment by greg0ire** [comment]
Yes, but that's not an imcompatible decision.

I think everything is summed up here: https://github.com/sonata-project/dev-kit/blob/68491f7e9af374586b76f6a6cc3460c6da7077cc/project/CONTRIBUTING.md#dependency-changes

**Comment by greg0ire** [comment]
@TimoBakx we are going to discuss that drop internally, don't bump that dependency in the meantime

File: src/Resources/config/core.xml (hunk 13 above)
Comment (TimoBakx) [reply src/Resources/config/core.xml]: Through the `Pool::getTemplate()` method, general templates are loaded (the ones that are not specific for an admin), like layout and ajax.

I could use `"sonata.admin.general_template_registry"` instead. I would prefer not to use "pool" anywhere.

Comment (greg0ire) [reply src/Resources/config/core.xml]: I think ajax is just another kind of layout, so maybe have that `layout` word in there?

File: src/Twig/Extension/TemplateRegistryExtension.php (hunk 15 above)
Comment (TimoBakx) [reply src/Twig/Extension/TemplateRegistryExtension.php]: I'd prefer to do this in a separate PR. I found another place that also needs it, so I could do both at the same time that way. Is that acceptable?

File: src/Twig/Extension/SonataAdminExtension.php (hunk 17 above)
Comment (TimoBakx) [reply src/Twig/Extension/SonataAdminExtension.php]: I was doubting about that as well.

I'll change it.

File: src/Resources/config/core.xml (hunk 13 above)
Comment (TimoBakx) [reply src/Resources/config/core.xml]: The templates `add_block`, `user_block` and `tab_menu_template` are also accessed this way.

Comment (greg0ire) [reply src/Resources/config/core.xml]: Ok, how about `global` instead? `general` feels a bit awkward?

Comment (TimoBakx) [reply src/Resources/config/core.xml]: Oh, that's better. Thanks!

File: src/Admin/TemplateRegistryInterface.php (hunk 3 above)
Comment (TimoBakx) [reply src/Admin/TemplateRegistryInterface.php]: I'm now using the constructor.

Would adding the new interface, plus the future servicelocator, justify a separate namespace, as we then have 4 classes/interfaces that belong together?

Comment (greg0ire) [reply src/Admin/TemplateRegistryInterface.php]: I think it would :+1: `Templating`?

Comment (TimoBakx) [reply src/Admin/TemplateRegistryInterface.php]: Inside `Admin`?

Comment (greg0ire) [reply src/Admin/TemplateRegistryInterface.php]: For now it's a component, but later it will be independent, so I'd rather say outside, but that's just my opinion.

Comment (TimoBakx) [reply src/Admin/TemplateRegistryInterface.php]: If I split up into a `TemplateRegistryInterface` and a `MutableTemplateRegistryInterface`, how should I do the typehinting? And should one interface extend the other?

Comment (greg0ire) [reply src/Admin/TemplateRegistryInterface.php]: I think `MutableTemplateRegistryInterface` should extend `TemplateRegistryInterface`, and the type hinting should be done depending on what the consuming class requires. For instance, the twig extensions will not need to set any template, will they?

Comment (TimoBakx) [reply src/Admin/TemplateRegistryInterface.php]: Alright, thanks. :-)

**Comment by Soullivaneuh** [comment]
Please see: https://github.com/sonata-project/dev-kit/issues/394#issuecomment-369169472

You can drop old version support of any dependency. The only rule is too keep at least one Symfony LTS support on both stable and master.

**Comment by TimoBakx** [comment]
Thanks for the feedback. I would still prefer to add the service locator and dropping support for Symfony DI <3.3 in a separate branch. I'd love to make that my next project after this PR is completed.

**Comment by greg0ire** [comment]
Sure, but be wary of the BC-breaks… will the type hint change if you inject a service locator instead of the entire container?

**Comment by TimoBakx** [comment]
I don't think so. The service locator uses ContainerInterface as well. I'll need to check if it moves from `Symfony\Component\DependencyInjection\ContainerInterface` to `Psr\Container\ContainerInterface`, though. But I'll make sure to test that in the new PR and change it if/when needed.
If it turns out to change, I'll target master instead of 3.x.

**Comment by TimoBakx** [comment]
I've updated the documentation.

File: docs/reference/templates.rst [hunk 18]
```diff
@@ -201,13 +202,21 @@ a global custom template and then override that customization on a specific
...
 ----------------------------
-Templates that are set using the ``setTemplate()`` or ``setTemplates()``
-methods can be accessed through the ``getTemplate($name)`` method of an
-Admin.
+Each ``Admin`` has a ``TemplateRegistry`` service connected to it that holds
+the templates registered through the configuration above. Through the method
+``getTemplate($name)`` of that class, you can access the templates set for
+that ``Admin``. The ``TemplateRegistry`` is available through ``$this->templateRegistry``
+within the ``Admin``. Using the service container the template registries can
+be accessed outside an ``Admin``. Use the ``Admin`` code + ".template_registry"
```
Comment (greg0ire) [inline docs/reference/templates.rst]: Use backticks for `template_registry` too please.

Comment (TimoBakx) [reply docs/reference/templates.rst]: Around the quotes or instead of the quotes?

Comment (greg0ire) [reply docs/reference/templates.rst]: :thinking: instead

File: src/Admin/AbstractAdmin.php [hunk 19]
```diff
@@ -1122,33 +1130,58 @@ public function generateMenuUrl($name, array $parameters = [], $absolute = Routi
         return $this->routeGenerator->generateMenuUrl($this, $name, $parameters, $absolute);
     }
 
+    public function setTemplateRegistry(MutableTemplateRegistryInterface $templateRegistry)
+    {
+        $this->templateRegistry = $templateRegistry;
+    }
+
+    /**
+     * @deprecated Use TemplateRegistry services instead
```
Comment (greg0ire) [inline src/Admin/AbstractAdmin.php]: If this is really deprecated, then you should trigger a deprecation. See the CONTRIBUTING.md guide for more details.

File: src/Admin/AbstractAdmin.php [hunk 20]
```diff
@@ -1122,33 +1130,58 @@ public function generateMenuUrl($name, array $parameters = [], $absolute = Routi
...
+     *
+     * @param array $templates
```
Comment (greg0ire) [inline src/Admin/AbstractAdmin.php]: Please remove this

File: src/Admin/Pool.php [hunk 21]
```diff
@@ -41,9 +42,9 @@ class Pool
     protected $adminClasses = [];
 
     /**
-     * @var string[]
+     * @var MutableTemplateRegistryInterface
      */
-    protected $templates = [];
+    protected $templateRegistry;
```
Comment (greg0ire) [inline src/Admin/Pool.php]: Should be private

File: src/Admin/Pool.php [hunk 22]
```diff
@@ -315,29 +316,39 @@ public function getAdminClasses()
         return $this->adminClasses;
     }
 
+    public function setTemplateRegistry(MutableTemplateRegistryInterface $templateRegistry)
```
Comment (greg0ire) [inline src/Admin/Pool.php]: Should be final

File: src/Admin/Pool.php (hunk 21 above)
Comment (TimoBakx) [reply src/Admin/Pool.php]: Should I add a `final protected getTemplateRegistry()` so people can access it in their admins?

Comment (greg0ire) [reply src/Admin/Pool.php]: No, if they need to, they will ask for it.

Comment (greg0ire) [reply src/Admin/Pool.php]: Plus this is `Pool`, so that wouldn't help, I guess, since `protected` would not be enough

Comment (TimoBakx) [reply src/Admin/Pool.php]: Oh, you're right.

File: src/Templating/MutableTemplateRegistryInterface.php [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Templating;
+
+interface MutableTemplateRegistryInterface extends TemplateRegistryInterface
+{
+    public function setTemplates(array $templates);
```
Comment (greg0ire) [inline src/Templating/MutableTemplateRegistryInterface.php]: Please document what the array should look like

File: src/Templating/TemplateRegistry.php [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Templating;
+
+final class TemplateRegistry implements TemplateRegistryInterface, MutableTemplateRegistryInterface
```
Comment (greg0ire) [inline src/Templating/TemplateRegistry.php]: One implements is enough because of inheritance

File: src/Admin/AbstractAdmin.php (hunk 19 above)
Comment (TimoBakx) [reply src/Admin/AbstractAdmin.php]: I'll remove the deprecation here until the admin specific configuration is implemented (see "next steps"). Until then, this is the only way to set the templates.

Comment (greg0ire) [reply src/Admin/AbstractAdmin.php]: Ah ok then.

**Comment by TimoBakx** [comment]
I added all requested changes.

**Comment by kunicmarko20** [comment]
Is this ready? (want to see that PR dropping < sf3.4) 😄 cc @sonata-project/contributors, please review

File: src/Admin/AbstractAdmin.php [hunk 25]
```diff
@@ -397,6 +398,8 @@
 
     /**
      * @var array
+     *
+     * @deprecated Use TemplateRegistry service instead
```
Comment (OskarStark) [inline src/Admin/AbstractAdmin.php]: @greg0ire is this message correct?

File: src/Admin/Pool.php [hunk 26]
```diff
@@ -40,11 +41,6 @@ class Pool
      */
     protected $adminClasses = [];
 
-    /**
-     * @var string[]
-     */
-    protected $templates = [];
```
Comment (OskarStark) [inline src/Admin/Pool.php]: this is a BC break

File: src/Templating/MutableTemplateRegistryInterface.php [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Templating;
+
+interface MutableTemplateRegistryInterface extends TemplateRegistryInterface
```
Comment (OskarStark) [inline src/Templating/MutableTemplateRegistryInterface.php]: please add `@author` tag

File: src/Templating/TemplateRegistry.php [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Templating;
+
+final class TemplateRegistry implements MutableTemplateRegistryInterface
```
Comment (OskarStark) [inline src/Templating/TemplateRegistry.php]: please add `@author` tag

File: src/Templating/TemplateRegistryInterface.php [hunk 29]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Templating;
+
+interface TemplateRegistryInterface
```
Comment (OskarStark) [inline src/Templating/TemplateRegistryInterface.php]: please add `@author` tag

File: src/Twig/Extension/SonataAdminExtension.php [hunk 30]
```diff
@@ -50,8 +54,17 @@ class SonataAdminExtension extends AbstractExtension
      */
     private $xEditableTypeMapping = [];
 
-    public function __construct(Pool $pool, LoggerInterface $logger = null, TranslatorInterface $translator = null)
-    {
+    /**
+     * @var ContainerInterface
```
Comment (OskarStark) [inline src/Twig/Extension/SonataAdminExtension.php]: looks wrong to me

File: src/Twig/Extension/SonataAdminExtension.php [hunk 31]
```diff
@@ -50,8 +54,17 @@ class SonataAdminExtension extends AbstractExtension
...
+     */
+    private $templateRegistries;
+
+    public function __construct(
+        Pool $pool,
+        LoggerInterface $logger = null,
+        TranslatorInterface $translator = null,
+        ContainerInterface $templateRegistries = null
```
Comment (OskarStark) [inline src/Twig/Extension/SonataAdminExtension.php]: is this type hint correct?

File: src/Admin/AbstractAdmin.php (hunk 25 above)
Comment (greg0ire) [reply src/Admin/AbstractAdmin.php]: I think so, what bothers you with it?

File: src/Twig/Extension/SonataAdminExtension.php (hunk 31 above)
Comment (TimoBakx) [reply src/Twig/Extension/SonataAdminExtension.php]: For now, I'll replace it with a ServiceLocator in a follow-up RP.

Comment (greg0ire) [reply src/Twig/Extension/SonataAdminExtension.php]: The service locator implements this interface, right?

Comment (TimoBakx) [reply src/Twig/Extension/SonataAdminExtension.php]: No, it implements `Psr\Container\ContainerInterface` instead of `Symfony\Component\DependencyInjection\ContainerInterface`.

Comment (TimoBakx) [reply src/Twig/Extension/SonataAdminExtension.php]: Fortunately, this is not confusing at all. ;-)

Comment (greg0ire) [reply src/Twig/Extension/SonataAdminExtension.php]: > When we drop support for <3.3, I can replace most references to Symfony\Component\DependencyInjection\ContainerInterface with Psr\Container\ContainerInterface.

And that will be a BC break :P

Comment (TimoBakx) [reply src/Twig/Extension/SonataAdminExtension.php]: Indeed. But that's a problem for another time.

Comment (kunicmarko20) [reply src/Twig/Extension/SonataAdminExtension.php]: But we said we will do this on stable as we don't have to support 2 LTS version (2.8 and 3.4)?

Comment (greg0ire) [reply src/Twig/Extension/SonataAdminExtension.php]: I think we did say that. So that would make it not a problem for another time, but a problem for now.

Comment (TimoBakx) [reply src/Twig/Extension/SonataAdminExtension.php]: So I should implement the Service Locator in this PR?

File: src/Admin/AbstractAdmin.php (hunk 25 above)
Comment (Soullivaneuh) [reply src/Admin/AbstractAdmin.php]: It is not. The deprecation should include "since 3.x, will be dropped in 4.0." first.

File: src/Twig/Extension/SonataAdminExtension.php (hunk 31 above)
Comment (Soullivaneuh) [reply src/Twig/Extension/SonataAdminExtension.php]: You can drop a version support, it's BC. The only rule is to drop it if the BC tricks are too heavy. Please see: https://github.com/sonata-project/SonataAdminBundle/blob/3.x/CONTRIBUTING.md#dependency-changes

Also, if BC tricks are in place concerning old dependency support, they should be removed as well.

Two options: On a separate commit or in a separate PR. The separate PR would be preferable because this one is already quite big.

Comment (TimoBakx) [reply src/Twig/Extension/SonataAdminExtension.php]: I also prefer a separate PR. Thank you for the extra info.

File: src/Admin/Pool.php (hunk 26 above)
Comment (TimoBakx) [reply src/Admin/Pool.php]: Can I solve this the same way as in the `AbstractAdmin` class (deprecating the attribute and keep setting it in the `setTemplates()` method)?

Comment (greg0ire) [reply src/Admin/Pool.php]: I would recommend doing so.

File: src/Twig/Extension/SonataAdminExtension.php (hunk 31 above)
Comment (TimoBakx) [reply src/Twig/Extension/SonataAdminExtension.php]: Maybe `$templateRegistryLocator` would be a better name to use with the `ContainerInterface`, but it's so long. :-(

Comment (kunicmarko20) [reply src/Twig/Extension/SonataAdminExtension.php]: It is descriptive? I don't mind long name, not sure what others think.

Comment (greg0ire) [reply src/Twig/Extension/SonataAdminExtension.php]: I was quite fine with `templateRegistries`, the fact that it's a container is kind of an implementation detail

**Comment by TimoBakx** [comment]
@OskarStark Could you please look over the review & replies again? Thank you.

**Comment by TimoBakx** [comment]
Rebasing removed all approvals. Please review again (the only conflict was in `UPGRADE-3.x.md`).

**Comment by TimoBakx** [comment]
Rebasing removed all approvals. Please review again (the only conflicts were in `UPGRADE-3.x.md` and `src/Resources/config/core.xml`).

**Comment by kunicmarko20** [comment]
Can we review and merge this before another rebase is needed?

**Comment by TimoBakx** [comment]
That's okay, @kunicmarko20. I can practise my command line GIT skills this way. ;-)

File: src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php [hunk 32]
```diff
@@ -382,11 +395,19 @@ public function fixTemplates(ContainerBuilder $container, Definition $definition
...
+            ->register($templateRegistryId, TemplateRegistry::class)
+            ->addTag('sonata.admin.template_registry')
+            ->setPublic(true); // Temporary fix until we can support service locators
+
         if ($container->getParameter('sonata.admin.configuration.templates') !== $definedTemplates) {
-            $definition->addMethodCall('setTemplates', [$definedTemplates]);
+            $templateRegistryDefinition->addArgument($definedTemplates);
         } else {
-            $definition->addMethodCall('setTemplates', ['%sonata.admin.configuration.templates%']);
+            $templateRegistryDefinition->addArgument('%sonata.admin.configuration.templates%');
```
Comment (jordisala1991) [inline src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: now that you touch this line, can you fix this other PR? it is on the same line

https://github.com/sonata-project/SonataAdminBundle/pull/4842

Can you check if with your code works and otherwise add what the other PR is suggesting?

File: src/Twig/Extension/TemplateRegistryExtension.php [hunk 33]
```diff
@@ -11,59 +11,82 @@
...
 
     public function getFunctions()
     {
         return [
             new TwigFunction('get_admin_template', [$this, 'getAdminTemplate']),
-            new TwigFunction('get_admin_pool_template', [$this, 'getPoolTemplate']),
+            new TwigFunction('get_global_template', [$this, 'getGlobalTemplate']),
+
+            // NEXT MAJOR: Remove this line
+            new TwigFunction('get_admin_pool_template', [$this, 'getGlobalTemplate'], ['deprecated' => true]),
```
Comment (jordisala1991) [inline src/Twig/Extension/TemplateRegistryExtension.php]: Shouldnt this be getPoolTemplate as before?

File: src/Twig/Extension/TemplateRegistryExtension.php [hunk 34]
```diff
@@ -11,59 +11,82 @@
...
-        return $this->getAdmin($adminCode)->getTemplate($name);
+        return $this->getTemplateRegistry($adminCode)->getTemplate($name);
     }
 
     /**
      * @param string $name
      *
      * @return null|string
      */
-    public function getPoolTemplate($name)
```
Comment (jordisala1991) [inline src/Twig/Extension/TemplateRegistryExtension.php]: isnt this a BC break?

File: src/Twig/Extension/TemplateRegistryExtension.php (hunk 33 above)
Comment (TimoBakx) [reply src/Twig/Extension/TemplateRegistryExtension.php]: I renamed the "pool" templates to "global" templates. They're the same list, just named differently.
They used to be stored inside the `Pool` class, that's why I named the function "get_admin_pool_template" at first. But "global" is a better name.

File: src/Twig/Extension/TemplateRegistryExtension.php (hunk 34 above)
Comment (TimoBakx) [reply src/Twig/Extension/TemplateRegistryExtension.php]: You're right. I'll put the original function back and deprecate it.

File: src/Twig/Extension/TemplateRegistryExtension.php (hunk 33 above)
Comment (jordisala1991) [reply src/Twig/Extension/TemplateRegistryExtension.php]: If you are restoring the old function, maybe you should restore the old call for this deprecated function. To preserve behavior

Comment (TimoBakx) [reply src/Twig/Extension/TemplateRegistryExtension.php]: Alright.

File: src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php [hunk 35]
```diff
@@ -382,11 +396,19 @@ public function fixTemplates(ContainerBuilder $container, Definition $definition
...
 
-        if ($container->getParameter('sonata.admin.configuration.templates') !== $definedTemplates) {
-            $definition->addMethodCall('setTemplates', [$definedTemplates]);
+        $templateRegistryId = $serviceId.'.template_registry';
+        $templateRegistryDefinition = $container
+            ->register($templateRegistryId, TemplateRegistry::class)
+            ->addTag('sonata.admin.template_registry')
+            ->setPublic(true); // Temporary fix until we can support service locators
+
+        if ($globalTemplates !== $definedTemplates) {
```
Comment (jordisala1991) [inline src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Now that I see this code..

if `globalTemplates` is different from `definedTemplates` you use `definedTemplates`, but if they are equal, you use `globalTemplates`...

Doesn't that mean that you always use definedTemplates? Or am I missing something?

Comment (TimoBakx) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Yeah, I'm starting to get more and more confused every time I look at this code. The original code uses the parameter by name instead of the actual array if the templates do not differ.
The PR you mentioned changes that to getting the value of that parameter instead of using the name.

In the end, it looks like that entire if-statement is not needed.

Comment (TimoBakx) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Running the code without the if-statement and without any `$globalTemplates`, the end result is exactly the same.

Comment (jordisala1991) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: https://github.com/sonata-project/SonataAdminBundle/pull/4580

This was done here... Maybe it has a good reason, but I don't know then why it does not work according to #4842 

Can you help us with this @nicolas-grekas ?

Comment (jordisala1991) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Seems like it is related to: https://github.com/symfony/symfony/pull/23143 

But that's only for SF 3.4, maybe older versions can't resolve that parameters.

Comment (nicolas-grekas) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Looking at the generated container shws the difference (at least at #4580 times): a reference to the parameter is dumped instead of its value. That was the goal.

File: src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php [hunk 36]
```diff
@@ -382,11 +395,15 @@ public function fixTemplates(ContainerBuilder $container, Definition $definition
...
-        } else {
-            $definition->addMethodCall('setTemplates', ['%sonata.admin.configuration.templates%']);
-        }
+        $templateRegistryId = $serviceId.'.template_registry';
+        $templateRegistryDefinition = $container
+            ->register($templateRegistryId, TemplateRegistry::class)
+            ->addTag('sonata.admin.template_registry')
+            ->setPublic(true); // Temporary fix until we can support service locators
+
+        $templateRegistryDefinition->addArgument($definedTemplates);
```
Comment (jordisala1991) [inline src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Forget what I said before, I think the other issue I mention might be addressed in another PR. Using the reference was on purpose. Sorry for the back and forth

Comment (TimoBakx) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: No problem. Shall I revert back?

Comment (jordisala1991) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Yes please

Comment (TimoBakx) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: And we're back. Long live `Undo`.

File: src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php (hunk 32 above)
Comment (TimoBakx) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Because this shows up in the PR main page now: We decided not to do this right now.

**Comment by kunicmarko20** [comment]
This should finally be done, let's review it for the last time (hopefully). cc @sonata-project/contributors

**Comment by TimoBakx** [comment]
Thank you all. :)
